compiler/aarch64/codegen/OMRMemoryReference.cpp

```diff
@@ -311,88 +319,97 @@ void OMR::ARM64::MemoryReference::setSymbol(TR::Symbol *symbol, TR::CodeGenerato
 
-      if (!node)
-         node = cg->getAppendInstruction()->getNode();
+         self()->setOffset(0);
 
-      if (_baseRegister != NULL)
-         {
-         if (constantIsUnsignedImm12(displacement))
+         if (_baseRegister && self()->isBaseModifiable())
+            newBase = _baseRegister;
+         else
             {
-            generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addimmx, node, newBase, _baseRegister, displacement);
+            newBase = cg->allocateRegister();
+
+            if (_baseRegister && _baseRegister->containsInternalPointer())
+               {
+               newBase->setContainsInternalPointer();
+               newBase->setPinningArrayPointer(_baseRegister->getPinningArrayPointer());
+               }
             }
-         else if (node->getOpCode().isLoadConst() && node->getRegister() && (node->getLongInt() == displacement))
+
+         if (_baseRegister != NULL)
             {
-            generateTrg1Src2Instruction(cg, TR::InstOpCode::addx, node, newBase, _baseRegister, node->getRegister());
+            if (constantIsUnsignedImm12(displacement))
+               {
+               generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addimmx, node, newBase, _baseRegister, displacement);
+               }
+            else if (node->getOpCode().isLoadConst() && node->getRegister() && (node->getLongInt() == displacement))
+               {
+               generateTrg1Src2Instruction(cg, TR::InstOpCode::addx, node, newBase, _baseRegister, node->getRegister());
+               }
+            else
+               {
+               TR::Register *tempReg = cg->allocateRegister();
+               loadConstant64(cg, node, displacement, tempReg);
+               generateTrg1Src2Instruction(cg, TR::InstOpCode::addx, node, newBase, _baseRegister, tempReg);
+               cg->stopUsingRegister(tempReg);
+               }
             }
          else
             {
-            TR::Register *tempReg = cg->allocateRegister();
-            loadConstant64(cg, node, displacement, tempReg);
-            generateTrg1Src2Instruction(cg, TR::InstOpCode::addx, node, newBase, _baseRegister, tempReg);
-            cg->stopUsingRegister(tempReg);
+            loadConstant64(cg, node, displacement, newBase);
             }
-         }
-      else
-         {
-         loadConstant64(cg, node, displacement, newBase);
-         }
 
-      // the following "if" is just to avoid stopUsingRegister being called
-      // for newBase by decNodeReferenceCounts(cg);
-      if (_baseRegister == newBase && _baseNode == NULL) _baseRegister = NULL;
+         if (_baseRegister != newBase)
+            {
+            self()->decNodeReferenceCounts(cg);
+            _baseNode = NULL;
+            self()->setBaseModifiable();
+            }
 
-      self()->decNodeReferenceCounts(cg);
-      _baseRegister = newBase;
-      _baseNode = NULL;
-      self()->setBaseModifiable();
+         _baseRegister = newBase;
```

Comment:
You can move this line into the `if (_baseRegister != newBase) { ... }` block above.

Reply:
Moved this line into the `if` block.